[mlir][py] Mark all type caster `from_{cpp,python}` methods as noexcept

nicholasjng opened the pull request:
This is mentioned as a "must" in https://nanobind.readthedocs.io/en/latest/porting.html#type-casters when implementing type casters.

While most of the existing `from_cpp` methods were already marked noexcept, many of the `from_python` methods were not. This commit adds the missing noexcept declarations to all type casters found in `NanobindAdaptors.h`.

nicholasjng committed:
[mlir][py] Mark all type caster `from_{cpp,python}` methods as noexcept

This is mentioned as a "must" in https://nanobind.readthedocs.io/en/latest/porting.html#type-casters when implementing type casters.

While most of the existing `from_cpp` methods were already marked noexcept, many of the `from_python` methods were not.
This commit adds the missing noexcept declarations to all type casters found in `NanobindAdaptors.h`.

ftynse reviewed (approved):
Thanks!

nicholasjng commented:
Hey, I'm still on this. 

My current idea is to return an invalid handle from `mlirApiObjectToCapsule` (i.e. one the wraps `nullptr`), but either the `mlirCapsuleTo*` family or the `PyCapsule_GetPointer` Python C API doesn't like that. I'm getting a `RuntimeError: std::bad_cast`, which I'm not sure how to address (I'll probably need to debug the bindings directly in C++).

Otherwise, my idea is to just return `nanobind::none` from `mlirApiObjectToCapsule`, and return early in `from_python` if `capsule.is_none()`. That is not as nice of a solution since it requires that logic to be inserted in all `from_python` type caster methods, and I'm not sure if the results play nice with the current testing assumptions.

I'll have to read up on the Python lit testing as well, since I don't know how to interpret the failed test output correctly (see below). But I'm working on it.

<details>
<summary>Test output</summary>

```
Command Output (stdout):
--
# RUN: at line 1
/Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/.venv/bin/python3 /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py pybind11 | /Users/nicholasjunge/Workspaces/c++/llvm-project/build/bin/FileCheck /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py
# executed command: /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/.venv/bin/python3 /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py pybind11
# executed command: /Users/nicholasjunge/Workspaces/c++/llvm-project/build/bin/FileCheck /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py
# RUN: at line 2
/Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/.venv/bin/python3 /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py nanobind | /Users/nicholasjunge/Workspaces/c++/llvm-project/build/bin/FileCheck /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py
# executed command: /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/.venv/bin/python3 /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py nanobind
# .---command stderr------------
# | Traceback (most recent call last):
# |   File "/Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py", line 329, in <module>
# |     @run
# |      ^^^
# |   File "/Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py", line 35, in run
# |     f()
# |   File "/Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py", line 362, in testCustomAttribute
# |     TestAttr(42)
# | RuntimeError: std::bad_cast
# `-----------------------------
# error: command failed with exit status: 1
# executed command: /Users/nicholasjunge/Workspaces/c++/llvm-project/build/bin/FileCheck /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py
# .---command stderr------------
# | /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py:424:16: error: CHECK-LABEL: expected string not found in input
# | # CHECK-LABEL: TEST: testTensorValue
# |                ^
# | <stdin>:50:26: note: scanning from here
# | TEST: testCustomAttribute
# |                          ^
# | <stdin>:51:3: note: possible intended match here
# | #python_test.test_attr
# |   ^
# | 
# | Input file: <stdin>
# | Check file: /Users/nicholasjunge/Workspaces/c++/llvm-project/mlir/test/python/dialects/python_test.py
# | 
# | -dump-input=help explains the following input dump.
# | 
# | Input was:
# | <<<<<<
# |              .
# |              .
# |              .
# |             45:  
# |             46: TEST: testOptionalOperandOp 
# |             47: op1.input is None: True 
# |             48: op2.input is None: False 
# |             49:  
# |             50: TEST: testCustomAttribute 
# | label:424'0                              X error: no match found
# |             51: #python_test.test_attr 
# | label:424'0     ~~~~~~~~~~~~~~~~~~~~~~~
# | label:424'1       ?                     possible intended match
# |             52: python_test.custom_attributed_op {test_attr = #python_test.test_attr} () -> () 
# | label:424'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |             53:  
# | label:424'0     ~
# |             54: #python_test.test_attr 
# | label:424'0     ~~~~~~~~~~~~~~~~~~~~~~~
# |             55: TestAttr(#python_test.test_attr) 
# | label:424'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | >>>>>>
# `-----------------------------
# error: command failed with exit status: 1

--

********************
********************
```

</details>

nicholasjng committed:
refactor: Change `mlirApiObjectToCapsule` to return std::optional

Instead of raising a `nanobind::type_error()`. This is necessary to honor the
nanobind type caster API contract, which requires `from_python` and `from_cpp` methods
to be marked `noexcept`.

nicholasjng commented:
Thanks, that's pretty cool. Although now I'm wondering why I didn't achieve the same with my previous `capsule.is_none()` checks, or with the current changeset.

Would you like to commit your version instead, then? Or how should we proceed?

makslevental commented:
> Would you like to commit your version instead, then? Or how should we proceed?

Just apply the patch your PR here - it's fine (I don't care "about" credit, just that it gets landed 😄)

nicholasjng committed:
refactor: Assign in if statements everywhere, adjust test expectations

Modeled after a patch posted in the conversation in PR #143866:
https://gist.github.com/makslevental/b224ffca7f15e273a4897975cda28b4c.

makslevental commented:
> The patch gives me a segmentation violation on my local machine (M1 Mac), though:

weird because i wrote and tested the patch on my m1... you're gonna have to build in debug mode to figure out where this is coming from

```
0   libMLIRPythonCAPI.dylib       	       0x16d5382a8 mlirAttributeIsAPythonTestTestAttribute + 8
1   _mlirPythonTestNanobind.cpython-312-darwin.so	       0x103ef40b8 0x103ef0000 + 16568
2   _mlirPythonTestNanobind.cpython-312-darwin.so	       0x103ef400c 0x103ef0000 + 16396
3   _mlirPythonTestNanobind.cpython-312-darwin.so	       0x103ef9d9c 0x103ef0000 + 40348
```

makslevental commented:
Lol okay it failed in CI - gotta whip a linux machine to test.

makslevental commented:
It's coming from here https://github.com/wjakob/nanobind/blob/master/include/nanobind/nb_cast.h#L576 which is a `std::bad_cast` https://github.com/wjakob/nanobind/blob/master/include/nanobind/nb_error.h#L90

makslevental commented:
But I think it's actually due to https://github.com/llvm/llvm-project/blob/244cf89f143a4fa09494280dba4a49d4cff007f6/mlir/include/mlir/Bindings/Python/NanobindAdaptors.h#L469

makslevental commented:
It's funny - what's happening is the cast is failing here 

```cpp
            MlirAttribute rawAttribute =
              nanobind::cast<MlirAttribute>(otherAttribute);
          if (!isaFunction(rawAttribute)) {
            auto origRepr =
                nanobind::cast<std::string>(nanobind::repr(otherAttribute));
            throw std::invalid_argument(
                (llvm::Twine("Cannot cast attribute to ") + captureTypeName +
                 " (from " + origRepr + ")")
                    .str());
          }
```

specifically `nanobind::cast<MlirAttribute>(otherAttribute)`. 

After stepping through a bunch I believe we're facing the fact that since `type_caster`s don't throw, we can't ourselves uses `nanobind::cast` without catching `std::bad_cast`... I feel like that's very annoying...

makslevental commented:
Yea this bears further investigation. I can't get to this until late next week so @nicholasjng feel free to keep pushing (but don't worry if you can't figure it out).

makslevental commented:
I guess the solution, if we do this, is to replace all the casts with [`try_cast`s](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I00EN8nanobind8try_castEbRKN6detail3apiI7DerivedEER1Tb)

makslevental commented:
here you go [this](https://gist.github.com/makslevental/034d262dfe703984825e95894ef0a2e4) works - i tested it [here](https://github.com/llvm/llvm-project/pull/148078) but you should land it here. we also need to check the rest of the casts to make sure if they're going through these type casters (if they're casting to `nb::cast<MlirSOMETHING>`) that either catch or get converted to `try_cast`.

nicholasjng committed:
Relax conversions to `nanobind::try_cast`

This avoids a nb::cast_error throw in certain class conversions.

makslevental commented:
> Thanks! I started checking the casts, and found two more `nanobind::cast<MlirTypeID>` in `lib/Bindings/Python/IRModule.h`:

Yea I saw these and I think they're safe according to exact your same logic. 

> It might also be necessary to check some of the other nanobind::cast(v) calls though, where v is some C++ type convertible to a MlirSOMETHING.

Yup those are exactly the ones I was worried about. Note though (I believe!) any of the casts like `nanobind::cast<PySOMETHING>()` should be safe because they do not go through an intermediate `MlirSOMETHING` step.

Also btw we're gonna need a PSA about this because it's a breaking change (even though it's a fix) for downstream projects.

nicholasjng commented:
Sure. Is there a process documentation for PSAs?

Since there are no other observable breakages with the current available tests, would you consider this otherwise complete for now?

nicholasjng commented:
That would be nice. I'll craft a PSA sometime later today. Thanks for all the help!

makslevental committed:
Merge branch 'main' into constexpr-from-cpp
